### PR TITLE
make image listing more resilient

### PR DESCRIPTION
**Pull request by vrothberg**
Handle more TOCTOUs operating on listed images.  Also pull in containers/common/pull/1520 which does the same on the internal layer tree.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2216700

#### Does this PR introduce a user-facing change?

```release-note
Make listing images more resilient towards concurrently running image removals.
```

### CODE REVIEW

**Comment by vrothberg** [comment]
Draft until https://github.com/containers/common/pull/1520 is merged.

**Comment by vrothberg** [comment]
@Luap99 @rhatdan PTAL

File: test/system/010-images.bats
Comment (Luap99) [inline test/system/010-images.bats:381]: At the very least we should store each pid then `wait` for it later and make sure all rmi's exited 0.

File: test/system/010-images.bats
Comment (Luap99) [inline test/system/010-images.bats:388]: What if one rmi fails, then this would loop forever , no?

**Review by Luap99 — COMMENTED** [review]
I would love for @edsantiago to have a look at the test.
code LGTM

File: test/system/010-images.bats
Comment (vrothberg) [reply test/system/010-images.bats:388]: Yes, that would loop forever.  I don't have an idea how to prevent that.  It would be easier to handle in the e2e tests but I really want to have a regression test in gating.

File: test/system/010-images.bats
Comment (vrothberg) [reply test/system/010-images.bats:381]: That's done at the end of the test.

File: test/system/010-images.bats
Comment (edsantiago) [inline test/system/010-images.bats:379]: debugging leftover

File: test/system/010-images.bats
Comment (edsantiago) [inline test/system/010-images.bats:388]: Suggestion:
```bash
tries=100
while [[ ${#lines[*]} -gt 1 ]] && [[ $tries -gt 0 ]]; do
    # Prior to #18980, 'podman images' during rmi could fail with 'image not known'
    run_podman images -q
    tries=$((tries - 1))
done
```

**Review by edsantiago — COMMENTED** [review]
Because this is such a freaky weird failure, I'd like to see a super-clear comment immediately next to the `podman images`. 

Suggested retry code is untested. Low on time today, heading out.

File: pkg/domain/infra/abi/images_list.go
Comment (TomSweeneyRedHat) [inline pkg/domain/infra/abi/images_list.go:40]: Only if you have other changes as it's an old nit
```suggestion
				return nil, fmt.Errorf("%w: ListData.IsDangling is nil but should not be", define.ErrInternal)
```

**Comment by TomSweeneyRedHat** [comment]
LGTM once the tests are hip and @edsantiago 's comments are addressed.

**Comment by vrothberg** [comment]
Need another PR in c/common.  I am pretty happy about the test as it's quite helpful in finding (more) issues.

**Comment by vrothberg** [comment]
Now vendoring https://github.com/containers/common/pull/1522 which is needed to address more errors that popped up.  I've been running the test for the past 20 mins on my machine but let's see what CI says.

File: test/system/010-images.bats
Comment (vrothberg) [reply test/system/010-images.bats:388]: Thanks!  Will update the PR once https://github.com/containers/common/pull/1522 has merged.

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by vrothberg** [comment]
Almost green and ready to merge

**Review by edsantiago — CHANGES_REQUESTED** [review]
/hold

File: test/system/010-images.bats
Comment (edsantiago) [inline test/system/010-images.bats:386]: I'm sorry, I wrote an abbreviated terse comment that was not actual code. I did not mean literal `--format Name`, because that is nonsensical. I just didn't want to type the curly braces dots quotes because I was in a hurry. I'm sorry.

Please change this to:
```bash
run_podman images --format '{{.ID}} {{.Names}}'
```
Reason: if the timeout test fails below, having both IID and image name will make it easier for a maintainer to understand what went wrong.

With this change, you can remove line 391 (`run_podman images`) below, because it does not contribute anything useful. The useful part of the error is "podman images spits out more than one line, and I was expecting only one line". The person investigating a test failure can easily see that already from all the repeated `podman images` calls in this loop.

I hope this makes sense. I am sorry for writing in shortcuts.

**Review by edsantiago — CHANGES_REQUESTED** [review]
Explanation.

File: test/system/010-images.bats
Comment (edsantiago) [reply test/system/010-images.bats:386]: Alternately, `run_podman images --noheading` might be even cleaner and easier for a maintainer to understand.

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
Flake turned out to be the unmount/EINVAL one (#18831), which is well known. This PR LGTM.

**Pull request merged**
